### ISSUE TITLE
Block [_id] from being ""

### ISSUE THREAD

**Issue by pickypg** [logstash-plugins/logstash-output-elasticsearch#587]
Related directly to: https://github.com/elastic/elasticsearch/pull/24118

Elasticsearch, starting with 5.3.1, will now block `_bulk` indexing requests with a value of `""`. This was always the case with ordinary, non-bulk indexing requests _and_ GET requests, so this was more of an unwanted exception that bulk requests allowed it.

Logstash should correctly identify and block such documents as well in the ES output to give a better reason (generally probably a malformed string).

**Comment by andrewvc** [comment]
@pickypg are we sure we want to block this on the logstash side? 

Elasticsearch will fail only that single document with the request right? Or, are you saying that the whole bulk request will fail? What's wrong with our regular logic here?

**Comment by pickypg** [comment]
> Elasticsearch will fail only that single document with the request right?

Correct; my phrasing was a bit dramatic, it will only block the specific indexing request that contains an `_id` of `""` with a `400` response.

There's nothing wrong with the regular logic, but specific error handling could better handle the scenario where I imagine someone doing something like

```ruby
elasticsearch {
   ...
   document_id => "%{my_var}"
}
```

where `my_var` happens to be `""` accidentally. If we detect it, we can log a better error that `_id` was `""` here rather than an otherwise hard to understand "bad request", which may not draw attention to the `_id`.

**Comment by andrewvc** [comment]
@pickypg that's a good point, though its unfortunate that ES doesn't return a great error. I agree that we should log a special error here.

**Comment by pickypg** [comment]
In fairness to LS, ES will spit out an error:

> if _id is specified it must not be empty

but I'm not honestly sure that this error would make sense when shown from Logstash since it's A) `document_id` and B) not really the _document_ rather metadata about it that's failing and I do not know if that's even shown.

**Comment by andrewvc** [comment]
Sadly, while fixing this is easy testing that fix is hard. I'm going to say we should defer work here until we refactor this plugin to make testing that logic easier.

**Comment by nicenemo** [comment]
I encountered this bug too.

## How to reproduce

Have an Elsatcisearch output config like:

```
output {
  elasticsearch {
    manage_template => false
    sniffing => false
    ssl => false
    hosts => ["192.168.23.42:9200"]
    ilm_enabled => true
    ilm_rollover_alias => "my-index"
    ilm_pattern => "000001"
    ilm_policy => "my-policy"
    doc_as_upsert => true
    action => "update"
    document_id => "%{myid}"
    http_compression => true
  }
}
```

Then have a bug in your ow software that sets the `myid` to empty string ,`""`.

What happens is that Logstash will retry over and over again and hog the CPU. The constant retries hog the CPU and  hold up output to other plugins in the same pipeline. On a memory constrained system it is almost impossible to stop the Logstash process except with `kill -9` or reboot.

## Workaround

Add a filter that drops the message when  `myid` is empty.

e.g. in Ruby(written by a Java programmer):

```Ruby
def session_id_exists(event)
  if event.get("myid").nil?
    return  false
  end

  if !event.get("myid").instance_of?(String)
    return false
  end

  if event.get("myid") == ""
    return false
  end
  return true
end
...

def filter(event)
...
   if ! session_id_exists(event)
    return []
  end
...
end
```

Also add output filtering to your own software.
